lib/deep_cover/autoload_tracker.rb

```diff
@@ -0,0 +1,208 @@
+          yield const, name, interceptor_path
+        end
+      end
+    end
+
+    # We need to remove the interceptor hooks, otherwise, the problem if manually requiring
+    # something that is autoloaded will cause issues.
+    def remove_interceptors(&do_autoload_block)
+      @autoloads_by_basename.each do |basename, entries|
+        entries.each do |entry|
+          const = entry.const
+          next unless const
+          # Module's constants are shared with Object. But if you set autoloads directly on Module, they
+          # appear on multiple classes. So just skip, Object will take care of those.
+          next if const == Module
+          if const.frozen?
+            warn_frozen_module(const)
+            next
+          end
+          yield const, entry.name, entry.target_path
+        end
+      end
+
+      @autoloaded_paths = {}
+      @interceptor_files_by_path = {}
+    end
+
+    protected
+
+    def setup_interceptor_for(const, name, path)
+      interceptor_path = autoload_interceptor_for(path)
+      entry = AutoloadEntry.new(WeakRef.new(const), name, path, interceptor_path)
+
+      basename = basename_without_extension(path)
+
+      @autoloads_by_basename[basename] ||= []
+      @autoloads_by_basename[basename] << entry
+      interceptor_path
+    end
+
+    def entries_for_target(requested_path, absolute_path_found)
+      basename = basename_without_extension(requested_path)
+      autoloads = @autoloads_by_basename[basename] || []
+
+      if absolute_path_found
+        autoloads.select { |entry| entry_is_target?(entry, requested_path, absolute_path_found) }
+      elsif requested_path == File.absolute_path(requested_path)
+        []
+      elsif requested_path.start_with?('./', '../')
+        []
+      else
+        # We didn't find a path that goes through the $LOAD_PATH
+        # It's possible that RubyGems will actually add the $LOAD_PATH and require an actual file
+        # So we must make a best-guest for possible matches
+        requested_path_to_compare = without_extension(requested_path)
+        autoloads.select { |entry| requested_path_to_compare == without_extension(entry.target_path) }
+      end
+    end
+
+    def entry_is_target?(entry, requested_path, absolute_path_found)
+      return true if entry.target_path == requested_path
+      target_path_rb = with_rb_extension(entry.target_path)
+      return true if target_path_rb == requested_path
+      entry_absolute_path = DeepCover.custom_requirer.resolve_path(entry.target_path)
+      return true if entry_absolute_path == absolute_path_found
+      false
+    end
+
+    def basename_without_extension(path)
+      without_extension(File.basename(path))
+    end
+
+    def with_rb_extension(path)
+      path += '.rb' unless needs_extension?(path)
+      path
+    end
+
+    def without_extension(path)
+      path = path[0...-3] unless needs_extension?(path)
+      path
+    end
+
+    def needs_extension?(path)
```

Comment:
Nitpicking: isn't `has_extention?` clearer than `needs_extension?`. If find need to be relative...

Reply:
"If find need to be relative...", not sure what you mean.

Well, `has_extention?` has the opposite meaning of `needs_extension?`. But yeah, this could use a better name: `has_supported_extension?`

Reply:
Some people say they "need" a Rolex. "need" is very relative in that case. Assessing "need" requires a metric of importance.
`has_supported_extension?` == :+1:

Reply:
Done